File: lib/core/device.dart
import 'dart:typed_data';

import 'package:rxdart/rxdart.dart';
import 'package:unisync/core/pairing_handler.dart';
import 'package:unisync/core/plugins/base_plugin.dart';
import 'package:unisync/core/plugins/clipboard/clipboard.plugin.dart';
import 'package:unisync/core/plugins/gallery/gallery.plugin.dart';
import 'package:unisync/core/plugins/notification/notification_plugin.dart';
import 'package:unisync/core/plugins/ring_phone/ring_phone.plugin.dart';
import 'package:unisync/core/plugins/run_command/run_command.plugin.dart';
import 'package:unisync/core/plugins/sharing/sharing.plugin.dart';
import 'package:unisync/core/plugins/status/status.plugin.dart';
import 'package:unisync/core/plugins/storage/storage.plugin.dart';
import 'package:unisync/core/plugins/telephony/telephony.plugin.dart';
import 'package:unisync/core/plugins/volume/volume.plugin.dart';
import 'package:unisync/utils/configs.dart';
import 'package:unisync/utils/logger.dart';
import 'package:unisync/utils/push_notification.dart';

import '../models/device_info/device_info.model.dart';
import '../models/device_message/device_message.model.dart';
import 'device_connection.dart';

class Device with ConnectionListener {
  factory Device(DeviceInfo info) {
    if (_instances.containsKey(info)) {
      return _instances[info]!;
    }
    _instances[info] = Device._(info);
    debugLog('Created an instance of Device.');
    _instanceNotifier.add(_instances.values.toList());
    debugLog('Device instances notified.');
    return _instances[info]!;
  }

  factory Device.fromId(String id) {
    return _instances[DeviceInfo(id: id, name: '', deviceType: '')]!;
  }

  Device._(this.info) {
    _initiate();
  }

  static final Map<DeviceInfo, Device> _instances = {};
  static final _instanceNotifier = BehaviorSubject<List<Device>>();

  static Stream<List<Device>> get instanceNotifier =>
      _instanceNotifier.asBroadcastStream();

  static void _removeInstance(DeviceInfo info) {
    final instance = _instances.remove(info);
    if (instance != null) {
      _instanceNotifier.add(_instances.values.toList());
    }
  }

  static Future<List<Device>> getAllDevices() async {
    for (final element in await ConfigUtil.device.getAllPairedDevices()) {
      Device(element);
    }
    return _instances.values.toList();
  }

  final DeviceInfo info;

  String? get ipAddress => _connection?.ipAddress;
  DeviceConnection? _connection;

  DeviceConnection? get connection => _connection;

  set connection(DeviceConnection? value) {
    _connection = value;
    if (value == null) {
      if (pairState != PairState.paired) {
        _removeInstance(info);
      }
      infoLog('Device@${info.name}: Disconnected!');
    } else {
      _connection!.connectionListener = this;
      infoLog('Device@${info.name}: Connected!');
      if (pairState == PairState.markUnpaired) {
        pairOperation.unpair();
      }
    }
    _notify();
  }

  late final _pairingHandler = PairingHandler(this, onStateChanged: (state) {
    _notify();
  });

  final List<UnisyncPlugin> plugins = [];

  final BehaviorSubject<DeviceNotification> _notifier = BehaviorSubject();

  Stream<DeviceNotification> get notifier => _notifier.asBroadcastStream();

  bool get isOnline => _pairingHandler.isReady && connection != null;

  PairState get pairState => _pairingHandler.state;

  PairOperation get pairOperation => _pairingHandler;

  Future<void> _initiate() async {
    await _pairingHandler.initiate();
  }

  @override
  void onDisconnected() {
    connection = null;
  }

  @override
  void onMessage(DeviceMessage message, Payload? payload) {
    infoLog('Device@${info.name}: Message received:');
    infoLog(message.toJson());
    if (message.type == DeviceMessage.Type.PAIR) {
      _pairingHandler.handle(message.body);
    } else if (_pairingHandler.state == PairState.paired) {
      for (final plugin in plugins) {
        if (plugin.type == message.type) {
          plugin.onReceive(message.header, message.body, payload);
        }
      }
    }
  }

  void sendMessage(DeviceMessage message, [Uint8List? data]) {
    if (_pairingHandler.state == PairState.paired ||
        message.type == DeviceMessage.Type.PAIR) {
      connection?.send(message, data);
      infoLog('Device@${info.name}: Message sent:');
      infoLog(message.toJson());
    }
  }

  T getPlugin<T extends UnisyncPlugin>() {
    return plugins.firstWhere((element) => element is T) as T;
  }

  void _initiatePlugins() {
    plugins.addAll([
      StatusPlugin(this),
      ClipboardPlugin(this),
      NotificationPlugin(this),
      VolumePlugin(this),
      RunCommandPlugin(this),
      RingPhonePlugin(this),
      TelephonyPlugin(this),
      SharingPlugin(this),
      GalleryPlugin(this),
      StoragePlugin(this),
    ]);
  }

  void _disposePlugins() {
    for (final plugin in plugins) {
      plugin.dispose();
    }
    plugins.clear();
  }

  void showNotification() {
    if (pairState == PairState.paired) {
      if (isOnline) {
        PushNotification.showNotification(
          title: 'Connection',
          text: 'Device ${info.name} connected!',
        );
      } else {
        PushNotification.showNotification(
          title: 'Connection',
          text: 'Device ${info.name} disconnected!',
        );
      }
    } else if (pairState == PairState.pairRequested) {
      PushNotification.showNotification(
        title: 'Pairing',
        text: 'Device ${info.name} requests to pair!',
      );
    }
  }

  void _notify() {
    _notifier.add(DeviceNotification(
      connected: isOnline,
      pairState: pairState,
    ));
    if (isOnline && pairState == PairState.paired) {
      _initiatePlugins();
    } else {
      _disposePlugins();
    }
  }
}

class DeviceNotification {
  DeviceNotification({
    this.connected = true,
    this.pairState = PairState.unknown,
  });

  final bool connected;
  final PairState pairState;
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/Device.kt
package com.anhquan.unisync.core

import android.content.Context
import com.anhquan.unisync.core.plugins.UnisyncPlugin
import com.anhquan.unisync.core.plugins.clipboard.ClipboardPlugin
import com.anhquan.unisync.core.plugins.gallery.GalleryPlugin
import com.anhquan.unisync.core.plugins.media.MediaPlugin
import com.anhquan.unisync.core.plugins.notification.NotificationPlugin
import com.anhquan.unisync.core.plugins.ring_phone.RingPhonePlugin
import com.anhquan.unisync.core.plugins.run_command.RunCommandPlugin
import com.anhquan.unisync.core.plugins.sharing.SharingPlugin
import com.anhquan.unisync.core.plugins.ssh.SSHPlugin
import com.anhquan.unisync.core.plugins.status.StatusPlugin
import com.anhquan.unisync.core.plugins.storage.StoragePlugin
import com.anhquan.unisync.core.plugins.telephony.TelephonyPlugin
import com.anhquan.unisync.core.plugins.volume.VolumePlugin
import com.anhquan.unisync.models.DeviceInfo
import com.anhquan.unisync.models.DeviceMessage
import com.anhquan.unisync.utils.ConfigUtil
import com.anhquan.unisync.utils.infoLog
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject

class Device private constructor(
    val context: Context, val info: DeviceInfo
) : DeviceConnection.ConnectionListener {
    companion object {
        private val instances = mutableMapOf<DeviceInfo, Device>()
        val instancesNotifier = BehaviorSubject.create<List<Device>>()

        fun of(context: Context, info: DeviceInfo): Device {
            if (instances.containsKey(info)) return instances[info]!!
            val device = Device(context, info)
            instances[info] = device
            notifyInstances()
            return device
        }

        fun of(deviceId: String): Device {
            return instances[DeviceInfo(id = deviceId, name = "", deviceType = "")]!!
        }

        private fun removeInstance(info: DeviceInfo) {
            val device = instances.remove(info)
            if (device != null) {
                notifyInstances()
            }
        }

        private fun notifyInstances() {
            instancesNotifier.onNext(instances.values.toList())
        }

        fun getAllDevices(context: Context): Single<List<Device>> {
            return ConfigUtil.Device.getAllPairedDevices().map {
                it.forEach { info -> of(context, info) }
                instances.values.toList()
            }
        }

        fun getConnectedDevices(): List<Device> {
            return instances.values.filter {
                it.isOnline
            }
        }
    }

    val ipAddress: String? get() = _connection?.ipAddress
    private var _connection: DeviceConnection? = null

    var connection: DeviceConnection?
        get() = _connection
        set(value) {
            if (_connection != null && value != null) return
            if (_connection == null && value == null) return
            _connection = value
            if (value == null) {
                infoLog("${this::class.simpleName}@${info.name}: Disconnected!")
                if (pairingHandler.state != PairingHandler.PairState.PAIRED) {
                    removeInstance(info)
                }
            } else {
                _connection!!.connectionListener = this
                infoLog("${this::class.simpleName}@${info.name}: Connected!")
                if (pairState == PairingHandler.PairState.MARK_UNPAIRED) {
                    pairOperation.unpair()
                }
            }
            notifyNewEvent()
        }

    val isOnline: Boolean get() = pairingHandler.isReady && connection != null
    val pairState: PairingHandler.PairState get() = pairingHandler.state
    val pairOperation: PairingHandler.PairOperation get() = pairingHandler.operation

    val plugins: MutableList<UnisyncPlugin> = mutableListOf()

    private val pairingHandler = PairingHandler(this) {
        notifyNewEvent()
    }

    override fun onMessage(message: DeviceMessage, payload: DeviceConnection.Payload?) {
        if (message.type == DeviceMessage.Type.PAIR) {
            pairingHandler.handle(message)
        } else if (pairingHandler.state == PairingHandler.PairState.PAIRED) {
            plugins.forEach { plugin ->
                if (plugin.type == message.type) {
                    plugin.onReceive(message, payload)
                }
            }
        }
    }

    override fun onDisconnected() {
        connection = null
    }

    fun sendMessage(
        message: DeviceMessage,
        payload: DeviceConnection.Payload? = null,
        onProgress: ((Float) -> Unit)? = null
    ) {
        if (pairingHandler.state == PairingHandler.PairState.PAIRED || message.type == DeviceMessage.Type.PAIR) {
            connection?.send(message, payload, onProgress)
            infoLog("${this::class.simpleName}@${info.name}: Message sent:\n$message")
        }
    }

    private fun initiatePlugins() {
        plugins.addAll(
            arrayOf(
                StatusPlugin(this),
                ClipboardPlugin(this),
                NotificationPlugin(this),
                VolumePlugin(this),
                RunCommandPlugin(this),
                RingPhonePlugin(this),
                TelephonyPlugin(this),
                SharingPlugin(this),
                GalleryPlugin(this),
                StoragePlugin(this),
                SSHPlugin(this),
                MediaPlugin(this)
            )
        )
    }

    private fun disposePlugins() {
        for (plugin in plugins) {
            plugin.onDispose()
        }
        plugins.clear()
    }

    fun <T : UnisyncPlugin> getPlugin(type: Class<T>): T {
        return plugins.filterIsInstance(type).first()
    }

    private val listeners = mutableListOf<DeviceEventListener>()

    fun addEventListener(listener: DeviceEventListener) {
        listeners.add(listener)
        listener.onDeviceEvent(
            DeviceEvent(
                connected = isOnline, pairState = pairState
            )
        )
    }

    fun removeEventListener(listener: DeviceEventListener) {
        listeners.remove(listener)
    }

    private fun notifyNewEvent() {
        if (isOnline && pairState == PairingHandler.PairState.PAIRED) {
            initiatePlugins()
            informListeners()
        } else {
            informListeners()
            disposePlugins()
        }
    }

    private fun informListeners() {
        notifyInstances()
        for (listener in listeners) {
            listener.onDeviceEvent(
                DeviceEvent(
                    connected = isOnline, pairState = pairState
                )
            )
        }
    }

    data class DeviceEvent(
        val connected: Boolean = true,
        val pairState: PairingHandler.PairState = PairingHandler.PairState.UNKNOWN
    )

    interface DeviceEventListener {
        fun onDeviceEvent(event: DeviceEvent)
    }
}